src/cryptonote_core/blockchain.cpp

```diff
@@ -1983,26 +1983,56 @@ void Blockchain::get_output_key_mask_unlocked(const uint64_t& amount, const uint
 //------------------------------------------------------------------
 bool Blockchain::get_output_distribution(uint64_t amount, uint64_t from_height, uint64_t to_height, uint64_t &start_height, std::vector<uint64_t> &distribution, uint64_t &base) const
 {
-  // rct outputs don't exist before v3
+  // rct outputs don't exist before v4
   if (amount == 0)
   {
     switch (m_nettype)
     {
-      case STAGENET: start_height = stagenet_hard_forks[2].height; break;
-      case TESTNET: start_height = testnet_hard_forks[2].height; break;
-      case MAINNET: start_height = mainnet_hard_forks[2].height; break;
+      case STAGENET: start_height = stagenet_hard_forks[3].height; break;
+      case TESTNET: start_height = testnet_hard_forks[3].height; break;
+      case MAINNET: start_height = mainnet_hard_forks[3].height; break;
       default: return false;
     }
   }
   else
     start_height = 0;
   base = 0;
 
+  if (to_height > 0 && to_height < from_height)
+    return false;
+
   const uint64_t real_start_height = start_height;
   if (from_height > start_height)
     start_height = from_height;
 
-  return m_db->get_output_distribution(amount, start_height, to_height, distribution, base);
+  distribution.clear();
+  uint64_t db_height = m_db->height();
+  if (db_height == 0)
+    return false;
+  if (to_height == 0)
+    to_height = db_height - 1;
+  if (start_height >= db_height || to_height >= db_height)
+    return false;
+  distribution.resize(to_height - start_height, 0);
+  if (amount == 0)
+  {
+    std::vector<uint64_t> heights;
+    heights.reserve(to_height + 1 - start_height);
+    uint64_t real_start_height = start_height > 0 ? start_height-1 : start_height;
+    for (uint64_t h = real_start_height; h <= to_height; ++h)
+      heights.push_back(h);
+    distribution = m_db->get_block_cumulative_rct_outputs(heights);
```

Comment:
I will at some point modify the RPC so I can pass single height values.

Reply:
Single height values meaning **one** or a `std::vector` of individual values?

The `std::vector<std::uint64_t>` argument into the LMDB class is difficult because if the typical case is a range, then LMDB should be used to jump to the first record and then iterated until the end of the range. If single and non-range based heights are provided over RPC, then LMDB should be used to jump to those individual heights.

Are two different functions into `db_lmdb.cpp` needed here?

Reply:
I was planning on a vector of values. Though that was mostly for genericity, I think the uses I have for it are either "all values" (gamma distribution picking) or "two values 1k apart" (segregation). AFAIK LMDB should do well whether requests are for consecutive records or not.